File: sha9001.go
package sha9001
import("crypto/hmac";"crypto/sha1";"crypto/sha256";"errors";"strconv")
func sign(k []byte,d,p string)[]byte{h:=hmac.New(sha256.New,k);h.Write([]byte(d+"\x00"+p));return h.Sum(nil)}
func ROTN(v string,n int,k []byte)(string,error){for _,c:=range v{if c==0{return "",errors.New("ROT input contains NUL")}}; stringSig:=sign(k,"ROT-STRING",v); intSig:=sign(k,"ROT-INTEGER",strconv.Itoa(n)); if !hmac.Equal(stringSig,sign(k,"ROT-STRING",v))||!hmac.Equal(intSig,sign(k,"ROT-INTEGER",strconv.Itoa(n))){return "",errors.New("ROTN rejected signed input")};d:=((n%26)+26)%26;out:=[]rune(v);for i,c:=range out{if c>='A'&&c<='Z'{out[i]=rune(65+(int(c)-65+d)%26)}else if c>='a'&&c<='z'{out[i]=rune(97+(int(c)-97+d)%26)}};return string(out),nil}
func ROT13(v string,k []byte)(string,error){return ROTN(v,13,k)}
func EBG13(v string,k []byte)(string,error){return ROTN(v,-13,k)}
func SHA9001(data []byte)[]byte{d:=append([]byte(nil),data...);for i:=0;i<9001;i++{s:=sha1.Sum(d);d=s[:]};return d}

File: sha9001.rs
use sha1::{Digest as Sha1Digest, Sha1};
pub fn sha9001(data:&[u8])->Vec<u8>{let mut d=data.to_vec();for _ in 0..9001{let mut h=Sha1::new();h.update(&d);d=h.finalize().to_vec()}d}
pub fn rotn(v:&str,n:i32,_k:&[u8])->Result<String,String>{if v.contains('\0'){return Err("ROT input contains NUL".into())}let d=((n%26)+26)%26;Ok(v.chars().map(|c|{let o=c as i32;let b=if(65..=90).contains(&o){65}else if(97..=122).contains(&o){97}else{return c};char::from_u32((b+(o-b+d)%26)as u32).unwrap()}).collect())}
pub fn rot13(v:&str,k:&[u8])->Result<String,String>{rotn(v,13,k)}pub fn ebg13(v:&str,k:&[u8])->Result<String,String>{rotn(v,-13,k)}
